kafka-rest/src/test/java/io/confluent/kafkarest/integration/v3/ConsumerLagsResourceIntegrationTest.java

```diff
@@ -77,67 +78,66 @@ public void setUp() throws Exception {
     produce(topic2, 1, request1);
 
-    // consume from subscribed topics
-    consumer1.poll(Duration.ofSeconds(1));
-    consumer2.poll(Duration.ofSeconds(1));
-    consumer1.commitSync();
-    consumer2.commitSync();
-
-    // stores expected currentOffsets and logEndOffsets for each topic partition after sending
-    // 3 records to topic1 partition0 and topic2 partition1
-    long[][] expectedOffsets = new long[numTopics][numPartitions];
-    expectedOffsets[0][0] = 3;
-    expectedOffsets[1][1] = 3;
-    // all other values default to 0L
-
-    Response response =
-        request("/v3/clusters/" + clusterId + "/consumer-groups/" + group1 + "/lags")
-            .accept(MediaType.APPLICATION_JSON)
-            .get();
+    KafkaConsumer<?, ?> consumer1 = createConsumer(group1, "client-1");
+    KafkaConsumer<?, ?> consumer2 = createConsumer(group1, "client-2");
 
-    assertEquals(Status.OK.getStatusCode(), response.getStatus());
-    ConsumerLagDataList consumerLagDataList =
-        response.readEntity(ListConsumerLagsResponse.class).getValue();
-
-    // checks offsets and lag match what is expected for each topic partition
-    for (int t = 0; t < numTopics; t++) {
-      for (int p = 0; p < numPartitions; p++) {
-        final int finalT = t;
-        final int finalP = p;
-        ConsumerLagData consumerLagData =
-            consumerLagDataList.getData()
-                .stream()
-                .filter(lagData ->
-                    lagData.getTopicName().equals(topics[finalT]))
-                .filter(lagData ->
-                    lagData.getPartitionId() == finalP)
-                .findAny()
-                .get();
-
-        assertEquals(expectedOffsets[t][p], (long) consumerLagData.getCurrentOffset());
-        assertEquals(expectedOffsets[t][p], (long) consumerLagData.getLogEndOffset());
-        assertEquals(0, (long) consumerLagData.getLag());
-      }
-    }
+    testWithRetry(
+        () -> {
+          consumer1.subscribe(Collections.singletonList(topic1));
+          consumer2.subscribe(Collections.singletonList(topic2));
+          consumer1.poll(Duration.ofSeconds(1));
+          consumer2.poll(Duration.ofSeconds(1));
+          // After polling once, only one of the consumers will be member of the group, so we poll again
+          // to force the other consumer to join the group.
+          consumer1.poll(Duration.ofSeconds(1));
+          consumer2.poll(Duration.ofSeconds(1));
+          // commit offsets from consuming from subscribed topics
+          consumer1.commitSync();
+          consumer2.commitSync();
+
+          // stores expected currentOffsets and logEndOffsets for each topic partition after sending
+          // 3 records to topic1 partition0 and topic2 partition1
+          long[][] expectedOffsets = new long[numTopics][numPartitions];
+          expectedOffsets[0][0] = 3;
+          expectedOffsets[1][1] = 3;
+          // all other values default to 0L
+
+          Response response =
```

Comment:
I think you are retrying too much code. Can't you just retry starting from here?

Reply:
I thought the test failures we saw (something along the lines of expected 3 but got 0 for getCurrentOffset() L135) stemmed from the consumer not polling/committing the offsets properly.

Reply:
I think there's a delay between whatever the consumer is doing above and the HTTP request below. The retry should only be in the HTTP request, to discard the possibility we just didn't wait long enough for the consumer to do its thing.